cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/upgrade/DataMigration26.java

```diff
@@ -0,0 +1,282 @@
+    if (runId != null) {
+      tagMap.put(Constants.Metrics.Tag.RUN_ID, runId);
+    }
+    Map<String, String> systemMap = null;
+    if (contextParts.size() > 0) {
+      systemMap = mapOldSystemContextToNew.get(contextParts.get(0));
+    }
+    String tagKey = metricNameToTagNameMapping.get(metricName);
+    if (systemMap != null) {
+      tagMap.putAll(systemMap);
+    } else  if (contextParts.size() > 1) {
+      // application metrics
+      List<String> targetTagList = typeToTagNameMapping.get(contextParts.get(1));
+      if (targetTagList != null) {
+        populateApplicationTags(tagMap, contextParts, targetTagList, context);
+      } else {
+        // service metrics , we are skipping them as they were not exposed for querying before
+        // todo : remove logging
+        LOG.info("Skipping context : {}", context);
+        return;
+      }
+    } else {
+      LOG.info("Should not reach here {}", context);
+    }
+    LOG.info("Calling Emit on the - tagMap : {} - context : {} - metricName : {} and tagKey : {}", tagMap, context,
+             metricName, tagKey);
+    emitMetrics(iterator, context, metricName, tagMap, tagKey);
+  }
+
+  private void populateApplicationTags(Map<String, String> tagMap, List<String> contextParts,
+                                       List<String> targetTagList, String context) {
+    tagMap.put(Constants.Metrics.Tag.NAMESPACE, Constants.DEFAULT_NAMESPACE);
+    for (int i = 0; i < contextParts.size(); i++) {
+      if (i == targetTagList.size()) {
+        LOG.info(" Context longer than targetTagList" + context);
+        break;
+      }
+      if (targetTagList.get(i).equals(TYPE)) {
+        continue;
+      }
+      tagMap.put(targetTagList.get(i), contextParts.get(i));
+    }
+  }
+
+  private void sendMetrics(MetricStore store, Map<String, String> tags, String metricName, int timeStamp,
+                           long value, MetricType counter) throws Exception {
+    LOG.info("Storing Metric - Tags {} metricName {} value {}", tags, metricName, value);
+    store.add(new MetricValue(tags, metricName, timeStamp, value, counter));
+  }
+
+  private MetricsTable getOrCreateMetricsTable(String tableName, DatasetProperties empty) {
+    LOG.info("Get Metrics Table" + tableName);
+    MetricsTable table = null;
+    // metrics tables are in the system namespace
+    Id.DatasetInstance metricsDatasetInstanceId = Id.DatasetInstance.from(Constants.SYSTEM_NAMESPACE, tableName);
+    while (table == null) {
+      try {
+        table = DatasetsUtil.getOrCreateDataset(dsFramework, metricsDatasetInstanceId,
+                                                MetricsTable.class.getName(), empty, null, null);
+      } catch (DatasetManagementException e) {
+        // dataset service may be not up yet
+        // todo: seems like this logic applies everywhere, so should we move it to DatasetsUtil?
+        LOG.warn("Cannot access or create table {}, will retry in 1 sec.", tableName);
```

Comment:
this was done to wait on dataset service to start. We don't really need it in "offline mode", so just fail here